libcodechecker/analyze/analysis_manager.py

```diff
@@ -408,14 +408,26 @@ def check(check_data):
                     for dependent_source in dependencies:
                         LOG.debug("[ZIP] Writing '" + dependent_source + "' "
                                   "to the archive.")
-                        archive_path = dependent_source.lstrip('/')
+                        archive_subpath = dependent_source.lstrip('/')
 
+                        archive_path = os.path.join('sources-root',
+                                                    archive_subpath)
                         try:
-                            archive.write(
-                                dependent_source,
-                                os.path.join("sources-root",
-                                             archive_path),
-                                zipfile.ZIP_DEFLATED)
+                            _ = archive.getinfo(archive_path)
```

Comment:
I wonder if this can be a bottleneck? An alternative approach would be to query the file list upfront and do not query the info for each path. In case this query is cheap, it should not be a problem.

Reply:
The `ZipFile` class (see `/usr/lib/python2.7/zipfile.py`) contains a `NameToInfo` dict which maps paths to `ZipInfo` objects. Dicts have `O(n)` worst-case lookup time.

However, this is currently a strange patch. If you look at the code, you can see we use `set`s everywhere for file lists, which also have a linear worst-case complexity. As explained earlier, I have no idea **why** some files are marked to be added multiple times to the ZIP. Perhaps something is wrong with parsing the analyzer output, perhaps this simply happens from @sylvestre using `--enable debug` checkers which produce *strange* output.

Reply:
And we are using string keys, so this is hopefully fast enough. We also aren't expecting **a lot** of analyses to fail. The ZIP file already goes over its internal structure at `.close()` (to finish up the ZIP structure, calculte CRCs and such), where it already figures out path duplications (hence the warning in #1219).